pkg/roachprod/vm/gce/gcloud.go

```diff
@@ -473,6 +481,225 @@ func (p *Provider) GetVMSpecs(
+		}
+	}
+	cacheKeys := buildCacheKeys(machineType, project, zone)
+
+	if cachedSpecs, ok := mts.cache[cacheKeys[ckl]]; ok {
+		return cachedSpecs, nil
+	}
+
+	// Query GCP to get machine type specs:
+	// Running:
+	// 		gcloud compute machine-types describe t2a-standard-4 \
+	//  		--project cockroach-ephemeral --zone us-east1-b \
+	// 			--format json
+	//
+	// Note: GCP requires a zone to describe a machine type, though during
+	// cluster creation, roachprod's takes a `--gce-machine-type` as input and
+	// guesses the CPU architecture from it. Only then, the zone is inferred
+	// from an hardcoded CPU architecture/zone mapping.
+	// As we're want to know the architecture from the machine type before
+	// knowing the zone, we try the default zone for both amd64 and arm64
+	// architectures. These calls are parallelized.
+	//
+	// In cases where we know the zone already (e.g. `roachprod list`), we use
+	// the supplied zone to query, making sure that we won't hit a zone in which
+	// the instance type is not available.
+
+	var zonesIteration []string
+	if zone != "" {
+		zonesIteration = []string{zone}
+	} else {
+		zonesIteration = []string{
+			DefaultZones(string(vm.ArchAMD64))[0],
+			DefaultZones(string(vm.ArchARM64))[0],
+		}
+	}
+
+	// gcloudQueryResult is the struct we use to pass around the query results.
+	// As we might iterate over multiple zones, we parallelize the calls and sink
+	// the results back via a channel.
+	type gcloudQueryResult struct {
+		specs *MachineTypeSpecs
+		err   error
+	}
+	resChan := make(chan gcloudQueryResult)
+	wg := sync.WaitGroup{}
+
+	for _, zone := range zonesIteration {
+		wg.Add(1)
+		go func(resChan chan<- gcloudQueryResult, machineType, project, zone string) {
+			args := []string{
+				"compute",
+				"machine-types",
+				"describe",
+				machineType,
+				"--project", project,
+				"--zone", zone,
+				"--format", "json",
+			}
+			specs := &MachineTypeSpecs{}
+			if err := runJSONCommand(args, &specs); err != nil {
+				if strings.Contains(err.Error(), "was not found") {
+					resChan <- gcloudQueryResult{
+						specs: nil,
+						err:   nil,
+					}
+					return
+				}
+				resChan <- gcloudQueryResult{
+					specs: nil,
+					err:   err,
+				}
+				return
+			}
+
+			resChan <- gcloudQueryResult{
+				specs: specs,
+				err:   nil,
+			}
+
+		}(resChan, machineType, project, zone)
+	}
+
+	// This goroutine sinks the gcloud query results into slices.
+	errs := []error{}
+	fetchedSpecs := []*MachineTypeSpecs{}
+	go func(resChan <-chan gcloudQueryResult) {
+		for res := range resChan {
+			if res.err != nil {
+				errs = append(errs, res.err)
+			}
+			if res.specs != nil {
+				fetchedSpecs = append(fetchedSpecs, res.specs)
+			}
+			wg.Done()
+		}
+	}(resChan)
+
+	// After waiting for all parallelized executions, we close the result channel,
+	// which will end the above sink results goroutine
+	wg.Wait()
+	close(resChan)
+
+	if len(errs) > 0 {
```

Comment:
What about the case `len(errs) == 1 && len(fetchedSpecs) == 1`? Since one of the queries is speculative, we shouldn't bail out just yet?